from modules.ui.controllers.BaseController import BaseController

from PySide6.QtCore import QCoreApplication as QCA

from modules.util.enum.GenerateMasksModel import GenerateMasksModel, GenerateMasksAction

from modules.ui.models.MaskModel import MaskModel

from modules.ui.utils.WorkerPool import WorkerPool

class MaskController(BaseController):
    def __init__(self, loader, parent=None):
        super().__init__(loader, "modules/ui/views/windows/generate_mask.ui", name=None, parent=parent)

    def _connectUIBehavior(self):
        self._connectFileDialog(self.ui.folderBtn, self.ui.folderLed, is_dir=True, save=False, title=
                               QCA.translate("dialog_window", "Open Dataset directory"))

        state_ui_connections = {
            "model": "modelCmb",
            "path": "folderLed",
            "prompt": "promptLed",
            "mode": "modeCmb",
            "alpha": "alphaSbx",
            "threshold": "thresholdSbx",
            "smooth": "smoothSbx",
            "expand": "expandSbx",
            "include_subdirectories": "includeSubfolderCbx"
        }

        self._connectStateUi(state_ui_connections, MaskModel.instance(), update_after_connect=True)

        self.__enableControls(True)()

        self.connect(self.ui.createMaskBtn.clicked, self.__startMask())


    def __createMask(self):
        def f(progress_fn=None):
            return MaskModel.instance().create_masks(progress_fn=progress_fn)

        return f

    def __startMask(self):
        def f():
            worker, name = WorkerPool.instance().createNamed(self.__createMask(), "create_mask", inject_progress_callback=True)
            if worker is not None:
                worker.connect(init_fn=self.__enableControls(False), result_fn=None,
                               finished_fn=self.__enableControls(True),
                               errored_fn=self.__enableControls(True), aborted_fn=self.__enableControls(True),
                               progress_fn=self._updateProgress(self.ui.progressBar))
                WorkerPool.instance().start(name)

        return f

    def __enableControls(self, enabled):
        def f():
            self.ui.createMaskBtn.setEnabled(enabled)
            if enabled:
                self.ui.progressBar.setValue(0)
        return f

    def _loadPresets(self):
        for e in GenerateMasksModel.enabled_values():
            self.ui.modelCmb.addItem(e.pretty_print(), userData=e)

        for e in GenerateMasksAction.enabled_values():
            self.ui.modeCmb.addItem(e.pretty_print(), userData=e)